### PR TITLE
builtins: Added pg_relation_is_updatable and pg_column_is_updatable builtins

### LINKED ISSUE #61715: sql: implement pg_relation_is_updatable and pg_column_is_updatable
Postgres has `pg_relation_is_updatable` and `pg_column_is_updatable` builtins.

I can't find these on the docs pages, but here they are in the code: https://github.com/postgres/postgres/blob/62f34097c88433ef1f3de604714fe7e7024f2fdf/src/backend/utils/adt/misc.c#L609

Tools like PostGraphile rely on this (https://github.com/graphile/postgraphile/issues/645). This should be easy to implement.

### PR DESCRIPTION
**Pull request by mnovelodou**
Previously, there were not functions pg_relation_is_updatable
and pg_column_is_updatable
This was inadequate because PostGraphile tool rely on these builtins
To address this, this patch adds these builtin functions

Release note (sql change): Added pg_relation_is_updatable and pg_column_is_updatable builtin functions

Fixes #61715

### CODE REVIEW

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 1]
```diff
@@ -83,6 +83,18 @@ var typeBuiltinsHaveUnderscore = map[oid.Oid]struct{}{
...
 
+// UpdateCommands Enumeration to match update operations in postgres
+const (
+	UpdateCommand = 2 + iota
+	InsertCommand
+	DeleteCommand
+)
+
+var (
+	zero = tree.NewDInt(0)
```
Comment (otan) [inline pkg/sql/sem/builtins/pg_builtins.go:94]: nit: zero is very vague in this case. i would do something like

```
type UpdatableCommand tree.DInt

const (
   nonUpdatableCommand UpdatableCommand = 0
   updateCommand UpdatableCommand = 2
// = 3, = 4 ...., all lowercase

   allUpdatableEvents = ( 1 << ... )
)
```

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 2]
```diff
@@ -1145,6 +1157,74 @@ SELECT description
...
+					return zero, nil
+				}
+				tableDesc, ok := table.(catalog.TableDescriptor)
+				if !ok || !tableDesc.IsTable() || tableDesc.IsVirtualTable() {
+					return zero, nil
+				}
+
+				return allEvents, nil
+			},
+			Info: "pg_relation_is_updatable was created for compatibility. This " +
```
Comment (otan) [inline pkg/sql/sem/builtins/pg_builtins.go:1179]: nit: you can avoid using `+` for strings by using ``` ` ```. i think we should omit the `allEvents` for now and leave it as a comment to allEvents above.

tl;dr - let's make the info:

```
Info: `"Returns the update events the relation supports.`
```

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 3]
```diff
@@ -1145,6 +1157,74 @@ SELECT description
...
+				}
+
+				column, err := tableDesc.FindColumnWithID(descpb.ColumnID(attNum))
+				if err != nil || column.IsComputed() {
+					return tree.DBoolFalse, nil
+				}
+
+				return tree.DBoolTrue, nil
+			},
+			Info: "pg_column_is_updatable was created for compatibility. This " +
```
Comment (otan) [inline pkg/sql/sem/builtins/pg_builtins.go:1221]: I would also make this Info the simplest thing possible:

```
Info: `Returns whether the given column can be updated.`
```

and leave the not virtual stuff as a comment above.

File: pkg/sql/sem/builtins/pg_builtins.go (hunk 1 above)
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go:94]: Done.

I kept nonUpdatableEvents and allUpdatableEvents as "variables" to have them as pointers of tree.DInt so I can return them as Datums.

File: pkg/sql/sem/builtins/pg_builtins.go (hunk 3 above)
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go:1221]: Done.

File: pkg/sql/logictest/testdata/logic_test/pg_builtins [hunk 4]
```diff
@@ -158,3 +158,68 @@ query TT
 SELECT pg_get_partkeydef(1), pg_get_partkeydef(NULL)
 ----
 NULL  NULL
+
+statement ok
+CREATE TABLE is_updatable(a INT PRIMARY KEY, b INT, c INT AS (b * 10) STORED);
+CREATE VIEW is_updatable_view AS SELECT a, b FROM is_updatable;
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/pg_builtins:164]: can you also remove the trailing ;?

**Review by otan — APPROVED** [review]
bors r+

nice!

**Review by otan — COMMENTED** [review]
found a few new things after your lint change!

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 5]
```diff
@@ -1145,6 +1160,79 @@ SELECT description
...
+		defProps(),
+		tree.Overload{
+			Types:      tree.ArgTypes{{"reloid", types.Oid}, {"include_triggers", types.Bool}},
+			ReturnType: tree.FixedReturnType(types.Int4),
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				oidArg := tree.MustBeDOid(args[0])
+				oid := int(oidArg.DInt)
+				table, err := ctx.Planner.GetImmutableTableInterfaceByID(ctx.Ctx(), oid)
+				if err != nil {
+					//nolint:returnerrcheck
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/sem/builtins/pg_builtins.go:1173]: postgres wasn't erroring for me:

```
postgres=# select pg_relation_is_updatable(-1, true);
 pg_relation_is_updatable
--------------------------
                        0
(1 row)

postgres=# select pg_relation_is_updatable(-1, false);
 pg_relation_is_updatable
--------------------------
                        0
(1 row)

postgres=# select pg_relation_is_updatable(231233321, false);
 pg_relation_is_updatable
--------------------------
                        0
(1 row)

postgres=# select pg_relation_is_updatable(231233321, true);
 pg_relation_is_updatable
--------------------------
                        0
(1 row)
```

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 6]
```diff
@@ -1145,6 +1160,79 @@ SELECT description
...
+				attNumArg := tree.MustBeDInt(args[1])
+				oid := int(oidArg.DInt)
+				attNum := uint32(attNumArg)
+				if attNumArg < 0 {
+					// system columns are not updatable
+					return tree.DBoolFalse, nil
+				}
+				table, err := ctx.Planner.GetImmutableTableInterfaceByID(ctx.Ctx(), oid)
+				if err != nil {
+					//nolint:returnerrcheck
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/sql/sem/builtins/pg_builtins.go:1212]: it didn't error for me

```
postgres=# select pg_column_is_updatable(123456, 2121::int2, true);
 pg_column_is_updatable
------------------------
 f
 ```

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 7]
```diff
@@ -1145,6 +1160,79 @@ SELECT description
...
+					return tree.DBoolFalse, nil
+				}
+				tableDesc, ok := table.(catalog.TableDescriptor)
+				if !ok || !tableDesc.IsTable() || tableDesc.IsVirtualTable() {
+					return tree.DBoolFalse, nil
+				}
+
+				column, err := tableDesc.FindColumnWithID(descpb.ColumnID(attNum))
+				if err != nil || column.IsComputed() {
+					//nolint:returnerrcheck
```
[the remark this answers is not in the document]
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go:1222]: Hi Oliver, let me check as I saw this on postgres code:
https://github.com/postgres/postgres/blob/master/src/backend/rewrite/rewriteHandler.c#L2754

`rel = try_relation_open(reloid, AccessShareLock);

	/*
	 * If the relation doesn't exist, return zero rather than throwing an
	 * error.  This is helpful since scanning an information_schema view under
	 * MVCC rules can result in referencing rels that have actually been
	 * deleted already.
	 */
	if (rel == NULL)
		return 0;`

[the remark this answers is not in the document]
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go:1222]: Look at this:

miguelnovelo=# SELECT pg_relation_is_updatable(1, true)
;
 pg_relation_is_updatable 
--------------------------
                        0
(1 row)

miguelnovelo=# SELECT count(*) FROM pg_class WHERE oid = 1;
 count 
-------
     0
(1 row)

miguelnovelo=#

[the remark this answers is not in the document]
Comment (otan) [reply pkg/sql/sem/builtins/pg_builtins.go:1228]:
```
otan=# \d bobby
                       Table "public.bobby"
 Column |          Type           | Collation | Nullable | Default
--------+-------------------------+-----------+----------+---------
 a      | geometry(Geometry,4326) |           |          |

otan=# select pg_column_is_updatable('bobby'::regclass::oid,55::int2,false);
 pg_column_is_updatable
------------------------
 t
(1 row)
```

if the column doesn't exist (55 certainly doesn't), looks like it returns true?

File: pkg/sql/sem/builtins/pg_builtins.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (otan) [reply pkg/sql/sem/builtins/pg_builtins.go:1218]: my comment about catching errors explicitly still holds here. maybe 
```
if sqlerrors.IsUndefinedTableError(err) {
   return true, nil
}
return false, err
}

[the remark this answers is not in the document]
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go]: Done.

File: pkg/sql/logictest/testdata/logic_test/pg_builtins [hunk 8]
```diff
@@ -158,3 +158,90 @@ query TT
...
+pg_class           relrowsecurity       4294967205  34      0                         false
+pg_class           relpartbound         4294967205  35      0                         false
+pg_class           relminmxid           4294967205  36      0                         false
+
+query B
+SELECT pg_column_is_updatable(oid, -1, true) FROM pg_class WHERE relname = 'is_updatable'
+----
+false
+
+# We make sure that the oid doesn't exists.
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/pg_builtins:227]: nit: "Check that the oid does not exist." (doesn't exists is a typo!)

File: pkg/sql/logictest/testdata/logic_test/pg_builtins [hunk 9]
```diff
@@ -158,3 +158,90 @@ query TT
...
+SELECT pg_relation_is_updatable(1, true)
+----
+0
+
+query B
+SELECT pg_column_is_updatable(1, 1, true)
+----
+false
+
+# is_updatable only have 3 columns
```
Comment (otan) [inline pkg/sql/logictest/testdata/logic_test/pg_builtins:243]: nit: "is_updatable only has 3 columns", but i'm not sure how this comment makes sense.
maybe we can make the things we're testing easy to read by doing something like this:
```
query TT
SELECT * FROM ( VALUES (
   ('system column', pg_column_is_updatable(oid, -1, true)),
   ('relation does not exist', pg_relation_is_updatable(1, true)),
   ('column on relation that does not exist', pg_column_is_updatable(1, 1, true)),
   ('column on relation that does exist', pg_column_is_updatable(oid, 15, true))
)) tbl(description, value)
```

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 10]
```diff
@@ -1145,6 +1161,91 @@ SELECT description
...
+				{"include_triggers", types.Bool},
+			},
+			ReturnType: tree.FixedReturnType(types.Bool),
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				oidArg := tree.MustBeDOid(args[0])
+				attNumArg := tree.MustBeDInt(args[1])
+				oid := int(oidArg.DInt)
+				attNum := uint32(attNumArg)
+				if attNumArg < 0 {
+					// system columns are not updatable
```
Comment (otan) [inline pkg/sql/sem/builtins/pg_builtins.go:1213]: nit: "System columns are not updatable."

**Review by otan — COMMENTED** [review]
i think you'll need to run crlfmt as well when the CI errors come about.

sorry for the back and forth!

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 11]
```diff
@@ -1145,6 +1161,91 @@ SELECT description
...
+				}
+				tableDesc, ok := table.(catalog.TableDescriptor)
+				if !ok || !tableDesc.IsTable() || tableDesc.IsVirtualTable() {
+					return tree.DBoolFalse, nil
+				}
+
+				column, err := tableDesc.FindColumnWithID(descpb.ColumnID(attNum))
+				if err != nil {
+					if sqlerrors.IsUndefinedColumnError(err) {
+						return tree.DBoolFalse, nil
```
Comment (otan) [inline pkg/sql/sem/builtins/pg_builtins.go:1234]: are you sure this is false? I get true. In PostgreSQL:

```
otan=# create table new_table(id int primary key);
CREATE TABLE
otan=# select pg_column_is_updatable('new_table'::regclass::oid, 3842::int2, false);
 pg_column_is_updatable
------------------------
 t
(1 row)
```

File: pkg/sql/logictest/testdata/logic_test/pg_builtins (hunk 8 above)
Comment (mnovelodou) [reply pkg/sql/logictest/testdata/logic_test/pg_builtins:227]: Done.

File: pkg/sql/logictest/testdata/logic_test/pg_builtins (hunk 9 above)
Comment (mnovelodou) [reply pkg/sql/logictest/testdata/logic_test/pg_builtins:243]: previous test: relation (and column) doesn't exist
this test: relation exists, column doesn't.

Let me change this, I didn't know that syntax of 'VALUES' thanks for the suggestion.

Comment (mnovelodou) [reply pkg/sql/logictest/testdata/logic_test/pg_builtins:243]: Done.

File: pkg/sql/sem/builtins/pg_builtins.go (hunk 11 above)
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go:1234]: Nice catch, you are right!

File: pkg/sql/sem/builtins/pg_builtins.go (hunk 10 above)
Comment (mnovelodou) [reply pkg/sql/sem/builtins/pg_builtins.go]: Done.

**Review by otan — COMMENTED** [review]
bors r+
